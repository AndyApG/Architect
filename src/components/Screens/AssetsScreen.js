import React from 'react';
import PropTypes from 'prop-types';
import Button from '@codaco/ui/lib/components/Button';
import Screen from '@components/Screen/Screen';
import AssetBrowser from '@components/AssetBrowser';
import SectionsContainer from '@components/sections/SectionsContainer';
import Section from '@components/sections/Section';
import SectionsHeading from '@components/sections/SectionsHeading';

const AssetBrowserScreen = ({
  show,
  transitionState,
  onComplete,
}) => {
  const buttons = [
    <Button
      key="done"
      onClick={onComplete}
      iconPosition="right"
    >
      Continue
    </Button>,
  ];

  return (
    <Screen
      show={show}
      buttons={buttons}
      transitionState={transitionState}
      onAcknowledgeError={onComplete}
    >
      <SectionsContainer>
        <SectionsHeading>
            Assets
        </SectionsHeading>
        <Section>
          <p>
            Welcome to the asset management screen. Here, you can load in images,
            video, audio, or even external network data which can be used elsewhere
            within your protocol.
          </p>
          <AssetBrowser />
        </Section>
      </SectionsContainer>
    </Screen>
  );
};

AssetBrowserScreen.propTypes = {
  show: PropTypes.bool,
  transitionState: PropTypes.string,
  onComplete: PropTypes.func.isRequired,
};

AssetBrowserScreen.defaultProps = {
  show: true,
  transitionState: null,
};

export { AssetBrowserScreen };

export default AssetBrowserScreen;
