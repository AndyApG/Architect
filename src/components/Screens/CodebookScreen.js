import React from 'react';
import PropTypes from 'prop-types';
import { Button } from '@codaco/ui/lib/components';
import Codebook from '@components/Codebook/Codebook';
import Screen from '@components/Screen/Screen';
import { Layout, Section } from '@components/EditorLayout';

/**
 * This component acts as an index for types. i.e. Nodes and Edges,
 * and links to the EditType.
 */
const CodebookScreen = (props) => {
  const {
    layoutId,
    onComplete,
    show,
    transitionState,
  } = props;

  const buttons = [
    <Button
      key="done"
      onClick={onComplete}
      iconPosition="right"
      icon="arrow-right"
    >
      Return to Timeline
    </Button>,
  ];

  return (
    <Screen
      show={show}
      buttons={buttons}
      transitionState={transitionState}
      onAcknowledgeError={onComplete}
      layoutId={layoutId}
    >
      <Layout>
        <Section title="Codebook">
          <p>
            Below you can find an overview of the node and edge types that you have
            defined while creating your interview. Entities that are unused may be deleted.
          </p>
        </Section>
        <Codebook />
      </Layout>
    </Screen>
  );
};

CodebookScreen.propTypes = {
  layoutId: PropTypes.string,
  onComplete: PropTypes.func,
  show: PropTypes.bool,
  transitionState: PropTypes.string,
};

CodebookScreen.defaultProps = {
  layoutId: null,
  onComplete: () => {},
  show: true,
  transitionState: null,
};

export default CodebookScreen;
