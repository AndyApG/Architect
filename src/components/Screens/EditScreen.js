import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { submit, isDirty } from 'redux-form';
import { Button } from '../../ui/components';
import Card from './ProtocolCard';

class EditScreen extends PureComponent {
  buttons() {
    const saveButton = (
      <Button
        key="save"
        onClick={this.props.submitForm}
        iconPosition="right"
      >
        Continue
      </Button>
    );

    return this.props.hasUnsavedChanges ? [saveButton] : [];
  }

  handleCancel = this.props.onComplete;

  render() {
    const {
      show,
      secondaryButtons,
      transitionState,
      editor: Editor,
      ...rest
    } = this.props;

    return (
      <Card
        buttons={this.buttons()}
        secondaryButtons={secondaryButtons}
        show={show}
        transitionState={transitionState}
        onCancel={this.handleCancel}
      >
        <Editor {...rest} />
      </Card>
    );
  }
}

EditScreen.defaultProps = {
  secondaryButtons: null,
};

const mapStateToProps = (state, { form }) => ({
  hasUnsavedChanges: isDirty(form)(state),
});

const mapDispatchToProps = (dispatch, { form }) => ({
  submitForm: () => dispatch(submit(form)),
});

export { EditScreen };

export default connect(mapStateToProps, mapDispatchToProps)(EditScreen);
