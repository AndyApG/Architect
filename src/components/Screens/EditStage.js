import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { compose } from 'recompose';
import {
  submit as submitForm,
  isDirty as isFormDirty,
  isInvalid as isFormInvalid,
} from 'redux-form';
import { has, find } from 'lodash';
import { Button } from '../../ui/components';
import Card from './ProtocolCard';
import StageEditor from '../../components/StageEditor';
import { getProtocol } from '../../selectors/protocol';
import { actionCreators as stageActions } from '../../ducks/modules/protocol/stages';
import { actionCreators as previewActions } from '../../ducks/modules/preview';

const formName = 'edit-stage';

class EditStage extends PureComponent {
  static propTypes = {
    dirty: PropTypes.bool.isRequired,
    invalid: PropTypes.bool.isRequired,
    show: PropTypes.bool.isRequired,
    submitForm: PropTypes.func.isRequired,
    onComplete: PropTypes.func,
    stage: PropTypes.object.isRequired,
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    insertAtIndex: PropTypes.number,
    updateStage: PropTypes.func.isRequired,
    createStage: PropTypes.func.isRequired,
    previewStage: PropTypes.func.isRequired,
    closePreview: PropTypes.func.isRequired,
    transitionState: PropTypes.string,
  };

  static defaultProps = {
    id: null,
    insertAtIndex: null,
    onComplete: () => {},
    transitionState: null,
  };

  get isDirty() {
    return this.props.dirty || !has(this.props.stage, 'id');
  }

  get buttons() {
    return this.isDirty ? [
      <Button
        key="continue"
        onClick={this.props.submitForm}
      >Continue</Button>,
    ] : [];
  }

  get secondaryButtons() {
    return [
      <Button
        key="preview"
        onClick={this.handlePreview}
        color="paradise-pink"
        disabled={this.props.invalid}
      >Preview</Button>,
    ];
  }

  handleComplete = () => {
    this.props.closePreview(); // Hide preview
    this.props.onComplete();
  }

  handlePreview = () => this.props.previewStage();

  handleSubmit = (stage) => {
    const { id, insertAtIndex } = this.props;

    if (id) {
      this.props.updateStage(id, stage);
    } else {
      this.props.createStage(stage, insertAtIndex);
    }

    this.handleComplete();
  }

  handleCancel = this.handleComplete;

  render() {
    const { stage, show, transitionState } = this.props;

    return (
      <Card
        buttons={this.buttons}
        secondaryButtons={this.secondaryButtons}
        show={show}
        transitionState={transitionState}
        onCancel={this.handleCancel}
      >
        <StageEditor
          stage={stage}
          form={formName}
          onSubmit={this.handleSubmit}
          previewStage={this.props.previewStage}
        />
      </Card>
    );
  }
}

const mapStateToProps = (state, props) => {
  const protocol = getProtocol(state);
  const stage = find(protocol.stages, ['id', props.id]) || { type: props.type };

  return ({
    stage,
    dirty: isFormDirty(formName)(state),
    invalid: isFormInvalid(formName)(state),
  });
};

const mapDispatchToProps = (dispatch, props) => {
  const stageMeta = {
    id: props.id,
  };

  return {
    submitForm: () => dispatch(submitForm(formName)),
    updateStage: bindActionCreators(stageActions.updateStage, dispatch),
    createStage: bindActionCreators(stageActions.createStage, dispatch),
    closePreview: bindActionCreators(previewActions.closePreview, dispatch),
    previewStage: () => dispatch(previewActions.previewStageByFormName(stageMeta, formName)),
  };
};

export { EditStage };

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
)(EditStage);
