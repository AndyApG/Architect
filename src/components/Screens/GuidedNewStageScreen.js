import React, { useState, useMemo } from 'react';
import { motion } from 'framer-motion';
import { Button } from '@codaco/ui';
import Screen from '@components/Screen/Screen';

const Steps = ({
  steps,
  initialStep,
}) => {
  const [history, setHistory] = useState([initialStep]);

  const previousStep = () => setHistory((s) => {
    const nextS = s.slice(0, -1);
    if (nextS.length === 0) { return [initialStep]; }
    return nextS;
  });

  const goToStep = (e) => {
    const step = e.target && e.target.getAttribute('data-step');
    if (!step) { return; }
    setHistory((s) => [...s, step]);
  };

  const currentStepId = useMemo(
    () => history[history.length - 1],
    [history],
  );

  const currentStep = steps.find(({ id }) => id === currentStepId);

  if (!currentStep) { return null; }

  const { actions, content } = currentStep;

  return (
    <motion.div className="guided-new-stage-screen__steps">
      <motion.div className="guided-new-stage-screen__step">
        <motion.div className="guided-new-stage-screen__step-content">
          {content}
        </motion.div>
        <motion.div className="guided-new-stage-screen__step-controls">
          <motion.div className="guided-new-stage-screen__step-back">
            { history.length > 1 && (
              <Button onClick={previousStep} color="charcoal">Previous step</Button>
            )}
          </motion.div>
          <motion.div className="guided-new-stage-screen__step-actions">
            { actions && actions.map(({
              label,
              step,
              color,
              onClick,
            }) => {
              const handleClick = step ? goToStep : onClick;
              return (
                <Button
                  onClick={handleClick}
                  data-step={step}
                  color={color}
                >
                  {label}
                </Button>
              );
            })}
          </motion.div>
        </motion.div>
      </motion.div>
    </motion.div>
  );
};

const steps = [
  {
    id: 'start',
    actions: [
      { label: 'Continue', step: 'ask-create-alters' },
    ],
    content: (
      <p>This wizard will ask you a series of questions.</p>
    ),
  },
  {
    id: 'ask-create-alters',
    actions: [
      { label: 'No', step: 'ask-create-edges', color: 'white' },
      { label: 'Yes', step: 'generator' },
    ],
    content: (
      <p>Do you want to create alters?</p>
    ),
  },
  {
    id: 'ask-create-edges',
    actions: [
      { label: 'No', step: 'ask-layout', color: 'white' },
      { label: 'Yes', step: 'sociogram' },
    ],
    content: (
      <p>Do you want to create edges?</p>
    ),
  },
  {
    id: 'ask-layout',
    actions: [
      { label: 'No', step: 'other', color: 'white' },
      { label: 'Yes', step: 'sociogram' },
    ],
    content: (
      <p>Do you want to create a layout of existing nodes?</p>
    ),
  },
  {
    id: 'generator',
    content: (<h1>Generator</h1>),
  },
  {
    id: 'sociogram',
    content: (<h1>Sociogram</h1>),
  },
  {
    id: 'other',
    content: (<h1>Other</h1>),
  }
];

const GuidedNewStageScreen = ({
  show,
  onSelect,
  onComplete,
}) => {
  const buttons = useMemo(() => [
    <Button
      key="cancel"
      onClick={onComplete}
      color="platinum"
    >
      Cancel
    </Button>,
  ], [onComplete]);

  return (
    <Screen
      show={show}
      buttons={buttons}
      type="guided-new-stage"
    >
      <motion.div className="guided-new-stage-screen">
        <div className="guided-new-stage-screen__section">
          <h1>Guided Stage Chooser</h1>
        </div>
        <Steps
          steps={steps}
          initialStep="start"
        />
      </motion.div>
    </Screen>
  );
};

export default GuidedNewStageScreen;
