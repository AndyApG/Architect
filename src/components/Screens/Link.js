import React, { useCallback } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { actionCreators as screenActions } from '@modules/ui/screens';

const Link = ({
  children,
  screen,
  openScreen,
  ...options
}) => {
  const handleOpenStage = useCallback(() => {
    openScreen(screen, options);
    onClick();
  }, ['openScreen']);

  return (
    <div onClick={handleOpenStage}>
      {children}
    </div>
  );
};

Link.propTypes = {
  openScreen: PropTypes.func.isRequired,
  screen: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
};

const mapDispatchToProps = {
  openScreen: screenActions.openScreen,
};

export default connect(null, mapDispatchToProps)(Link);
