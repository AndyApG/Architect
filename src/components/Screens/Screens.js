import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { compose } from 'recompose';
import { motion } from 'framer-motion';
import { get } from 'lodash';
import Window from '@codaco/ui/lib/components/window';
import { getCSSVariableAsNumber } from '@codaco/ui/lib/utils/CSSVariables';
import { getScreensStack } from '@selectors/ui';
import { actionCreators as uiActions } from '@modules/ui';
import { getScreenComponent } from './screenIndex';

/**
 * Screen manager for Architect.
 *
 * Any screens present in the `screens` redux state will be rendered by this
 * component.
 *
 * Screens can be opened and closed using the screen module actions:
 * - `openScreen(name, params)`:
 *    `name` is used to find the screen component (screens are listed by name in screenIndex),
 *     and `params` are passed to screen component itself
 *    This will be rendered as `<ScreenComponent {...params} />`
 * - `closeScreen(name, params)`
 */
const Screens = (props) => {
  const screens = props.screens.map(({ screen, params }, index) => {
    const ScreenComponent = getScreenComponent(screen);
    const onComplete = (result) => props.closeScreen(screen, result);

    const zIndex = getCSSVariableAsNumber('--z-panel') + index;

    return (
      <ScreenComponent
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...params}
        zIndex={zIndex}
        layoutId={params.id}
        onComplete={onComplete}
      />
    );
  });

  return screens;
};

Screens.propTypes = {
// eslint-disable-next-line react/forbid-prop-types
screens: PropTypes.array.isRequired,
closeScreen: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
screens: getScreensStack(state),
});

const mapDispatchToProps = {
closeScreen: uiActions.closeScreen,
};

export { Screens };

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  Window,
)(Screens);
