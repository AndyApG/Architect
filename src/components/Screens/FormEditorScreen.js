import { compose, withProps } from 'recompose';
import FormEditor, { formName } from '../FormEditor';
import EditScreen from './EditScreen';

const formEditorProps = withProps({
  editor: FormEditor,
  form: formName,
});

const FormEditorScreen = compose(
  formEditorProps,
)(EditScreen);

export default FormEditorScreen;
