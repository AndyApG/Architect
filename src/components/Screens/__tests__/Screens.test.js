/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import Screens from '../Screens';

const mockProps = {
  screens: [],
  closeScreen: () => {},
};

describe('<Screens />', () => {
  it('can render?', () => {
    const component = shallow(<Screens {...mockProps} />);

    expect(component).toMatchSnapshot();
  });
});
