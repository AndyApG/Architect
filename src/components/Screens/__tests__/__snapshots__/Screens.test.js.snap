// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Screens /> can render? 1`] = `
<TransitionGroup
  childFactory={[Function]}
  className="screens"
  component="div"
/>
`;
